File: app/src/main/java/com/oracleairescue/DiffUtil.java
package com.oracleairescue;

class DiffUtil {
    static String stripCodeFence(String text) {
        if (text == null) return "";
        String t = text.trim();
        if (t.startsWith("```")) {
            int firstNewline = t.indexOf('\n');
            int lastFence = t.lastIndexOf("```");
            if (firstNewline >= 0 && lastFence > firstNewline) t = t.substring(firstNewline + 1, lastFence).trim();
        }
        return t;
    }

    static String unifiedDiff(String oldText, String newText) {
        String[] a = oldText == null ? new String[0] : oldText.split("\\R", -1);
        String[] b = newText == null ? new String[0] : newText.split("\\R", -1);
        StringBuilder sb = new StringBuilder();
        sb.append("--- original\n+++ modified\n");
        int max = Math.max(a.length, b.length);
        int shown = 0;
        for (int i = 0; i < max; i++) {
            String left = i < a.length ? a[i] : null;
            String right = i < b.length ? b[i] : null;
            if (left == null) { sb.append("+").append(right).append("\n"); shown++; }
            else if (right == null) { sb.append("-").append(left).append("\n"); shown++; }
            else if (!left.equals(right)) {
                sb.append("-").append(left).append("\n");
                sb.append("+").append(right).append("\n");
                shown++;
            }
            if (shown > 2000) { sb.append("...diff 過長，已截斷...\n"); break; }
        }
        if (shown == 0) sb.append("沒有差異。\n");
        return sb.toString();
    }
}

File: app/src/main/java/com/oracleairescue/DiffUtil.kt
package com.oracleairescue

object DiffUtil {
    fun unifiedDiff(oldText: String, newText: String, maxLines: Int = 1200): String {
        val oldLines = oldText.split("\n")
        val newLines = newText.split("\n")
        val out = StringBuilder()
        out.appendLine("--- 原檔案")
        out.appendLine("+++ 修改後")
        var i = 0
        var j = 0
        var shown = 0
        while ((i < oldLines.size || j < newLines.size) && shown < maxLines) {
            val oldLine = oldLines.getOrNull(i)
            val newLine = newLines.getOrNull(j)
            when {
                oldLine == newLine -> {
                    if (oldLine != null) {
                        out.appendLine("  $oldLine")
                        i++; j++; shown++
                    } else break
                }
                newLine != null && oldLines.drop(i + 1).take(6).contains(newLine) -> {
                    out.appendLine("- ${oldLine.orEmpty()}")
                    i++; shown++
                }
                oldLine != null && newLines.drop(j + 1).take(6).contains(oldLine) -> {
                    out.appendLine("+ ${newLine.orEmpty()}")
                    j++; shown++
                }
                oldLine != null && newLine != null -> {
                    out.appendLine("- $oldLine")
                    out.appendLine("+ $newLine")
                    i++; j++; shown += 2
                }
                oldLine != null -> {
                    out.appendLine("- $oldLine")
                    i++; shown++
                }
                newLine != null -> {
                    out.appendLine("+ $newLine")
                    j++; shown++
                }
            }
        }
        if (i < oldLines.size || j < newLines.size) out.appendLine("... diff 過長，已截斷 ...")
        return out.toString()
    }

    fun stripCodeFence(text: String): String {
        val trimmed = text.trim()
        if (!trimmed.startsWith("```")) return text.trimEnd()
        val lines = trimmed.lines().toMutableList()
        if (lines.isNotEmpty() && lines.first().startsWith("```")) lines.removeAt(0)
        if (lines.isNotEmpty() && lines.last().startsWith("```")) lines.removeAt(lines.lastIndex)
        return lines.joinToString("\n").trimEnd()
    }
}
